artemis-server/src/main/java/org/apache/activemq/artemis/core/persistence/impl/journal/LargeServerMessageImpl.java

```diff
@@ -368,52 +368,48 @@ public Message copy(final long newID) {
+         //clone a SequentialFile to avoid concurrent access
+         ensureFileExists(false);
+         SequentialFile cloneFile = file.cloneFile();
 
-         validateFile();
-
-         byte[] bufferBytes = new byte[100 * 1024];
-
-         ByteBuffer buffer = ByteBuffer.wrap(bufferBytes);
+         try {
+            byte[] bufferBytes = new byte[100 * 1024];
 
-         long oldPosition = file.position();
+            ByteBuffer buffer = ByteBuffer.wrap(bufferBytes);
 
-         if (!file.isOpen()) {
-            file.open();
-         }
-         file.position(0);
-
-         for (;;) {
-            // The buffer is reused...
-            // We need to make sure we clear the limits and the buffer before reusing it
-            buffer.clear();
-            int bytesRead = file.read(buffer);
-
-            byte[] bufferToWrite;
-            if (bytesRead <= 0) {
-               break;
-            } else if (bytesRead == bufferBytes.length && !this.storageManager.isReplicated()) {
-               // ARTEMIS-1220: We cannot reuse the same buffer if it's replicated
-               // otherwise there could be another thread still using the buffer on a
-               // replication.
-               bufferToWrite = bufferBytes;
-            } else {
-               bufferToWrite = new byte[bytesRead];
-               System.arraycopy(bufferBytes, 0, bufferToWrite, 0, bytesRead);
+            if (!cloneFile.isOpen()) {
+               cloneFile.open();
             }
 
-            newMessage.addBytes(bufferToWrite);
-
-            if (bytesRead < bufferBytes.length) {
-               break;
+            cloneFile.position(0);
+
+            for (;;) {
+               // The buffer is reused...
+               // We need to make sure we clear the limits and the buffer before reusing it
+               buffer.clear();
+               int bytesRead = cloneFile.read(buffer);
+
+               byte[] bufferToWrite;
+               if (bytesRead <= 0) {
+                  break;
+               } else if (bytesRead == bufferBytes.length && !this.storageManager.isReplicated()) {
+                  // ARTEMIS-1220: We cannot reuse the same buffer if it's replicated
+                  // otherwise there could be another thread still using the buffer on a
+                  // replication.
+                  bufferToWrite = bufferBytes;
+               } else {
+                  bufferToWrite = new byte[bytesRead];
+                  System.arraycopy(bufferBytes, 0, bufferToWrite, 0, bytesRead);
+               }
+
+               newMessage.addBytes(bufferToWrite);
+
+               if (bytesRead < bufferBytes.length) {
+                  break;
+               }
             }
-         }
-
-         file.position(oldPosition);
-
-         if (!originallyOpen) {
-            file.close(false);
-            newMessage.getFile().close();
+         } finally {
+            cloneFile.close();
          }
```

Comment:
Do we need to close underlying file of new large message after copy?

Reply:
yes it should be closed as original code. That'll cause file leak.

Reply:
@clebertsuconic @gaohoward I don't understand why we only close file when old file is not originally opened? In your case where other consumers open file and deliver, the file of new large message might not be closed. This will result in file leak and maybe data corrupt if broker crashes, wdyt?

Reply:
tbh I don't know either. I just keep the old behavior.

Reply:
@wy96f I'm doing some work with large messages, and I want to double check that as well.

Reply:
@wy96f  / @gaohoward actually.. I will merge this right away.. as whatever I do would clash here.


and I will make sure I address how we open and close files.